### PR TITLE
Turn on strict dtype checking for test_torch.py

### LINKED ISSUE #33821: Method overrides and instantiate_device_type_tests don't play nicely
Steps to reproduce:
1. Add an override to `assertEqual` in a class that is instantiated by `instantiate_device_type_tests`
2. Run the test and see if your override is called

Expected result: It is called
Actual result: it is not

cc @mruberry

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33826 Stop filling in defaults in Python binding layer.
* #33518 Collapse _like overloads into a single overload.
* #33516 Add MemoryFormat to TensorOptions, but not codegen.
* **#33825 Turn on strict dtype checking for test_torch.py**

Partially addresses #20376

I do this by overriding assertEqual in classes that opt into
this.  This means I have to fix #33821.  The fix is a little
unsatisfactory as idiomatic Python 2 super() calls don't work
(since the class is no longer in scope); hopefully this will just
work when we go to Python 3.

General approach taken:
- A lot of dtype mismatches are because we specified tensor constants
  that infer to some dtype, but the actual dtype needed is something else.
  Those are easy, just annotate the tensor() constructor (often a legacy
  Tensor/FloatTensor call) with dtype
- There are a few cases where the promotion rules are nontrivial.  Some of them
  I just typed out the expected promotion rules manually (based on trial
  and error)
- There are some more complex cases; if it gets too hairy I just
  set exact_dtype=False and nope the fuck out

I don't have time to do it for all the other classes.  But the setup
should work if people just incrementally add the overrides to classes,
and then eventually flip the default.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20125791](https://our.internmc.facebook.com/intern/diff/D20125791)

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -15301,7 +15354,9 @@ def test_svd_tall_all_col_maj(self, device, dtype):
         self._test_svd_helper((5, 20), False, True, device, dtype)
 
 class TestTorch(TestCase, _TestTorchMixin):
-    pass
+    def assertEqual(self, *args, **kwargs):
+        kwargs.setdefault('exact_dtype', True)
+        super(TestTorch, self).assertEqual(self, *args, **kwargs)
```
Comment (mruberry) [inline test/test_torch.py]: This version of the function is slightly different than the others. I suppose you could make a class that inherits from TestCase, "StrictTestCase," which just overrides assertEqual as the above classes do in your current PR, and then inherit these test classes from it? That might be more maintainable.

**Review by mruberry — COMMENTED** [review]
Your approach is excellent, you've solved problems you didn't even intend to and your code is beautiful as always, @ezyang, but you still have test failures. Ping me again when they're addressed?

File: test/test_torch.py (hunk 1 above)
Comment (ezyang) [reply test/test_torch.py]: This is a great idea. In fact, I don't even need a subclass, I can just do the toggle on a class member

File: torch/testing/_internal/common_device_type.py [hunk 2]
```diff
@@ -270,7 +270,7 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None):
     empty_class = type(empty_name, generic_test_class.__bases__, {})
 
     # Acquires members names
-    generic_members = set(dir(generic_test_class)) - set(dir(empty_class))
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/testing/_internal/common_device_type.py]: done

**Comment by ezyang** [comment]
Needs XLA fix https://github.com/pytorch/xla/pull/1695

**Comment by ezyang** [comment]
@mruberry tests are all green now PHEW

**Review by mruberry — APPROVED** [review]
LGTM!

**Comment by VitalyFedyunin** [comment]
cc @VitalyFedyunin to track landing status

**Pull request closed without merging**